connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Flatten.java

```diff
@@ -0,0 +1,233 @@
+import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
+
+public abstract class Flatten<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    public static final String OVERVIEW_DOC =
+            "Flatten a nested data structure, generating names for each field by concatenating the field names at each "
+                    + "level with a configurable delimiter character. Applies to Struct when schema present, or a Map "
+                    + "in the case of schemaless data. The default delimiter is '.'."
+                    + "<p/>Use the concrete transformation type designed for the record key (<code>" + Key.class.getCanonicalName() + "</code>) "
+                    + "or value (<code>" + Value.class.getCanonicalName() + "</code>).";
+
+    private static final String DELIMITER_CONFIG = "delimiter";
+    private static final String DELIMITER_DEFAULT = ".";
+
+    public static final ConfigDef CONFIG_DEF = new ConfigDef()
+            .define(DELIMITER_CONFIG, ConfigDef.Type.STRING, DELIMITER_DEFAULT, ConfigDef.Importance.MEDIUM,
+                    "Delimiter to insert between field names from the input record when generating field names for the "
+                            + "output record");
+
+    private static final String PURPOSE = "flattening";
+
+    private String delimiter;
+
+    private Cache<Schema, Schema> schemaUpdateCache;
+
+    @Override
+    public void configure(Map<String, ?> props) {
+        final SimpleConfig config = new SimpleConfig(CONFIG_DEF, props);
+        delimiter = config.getString(DELIMITER_CONFIG);
+        schemaUpdateCache = new SynchronizedCache<>(new LRUCache<Schema, Schema>(16));
+    }
+
+    @Override
+    public R apply(R record) {
+        if (operatingSchema(record) == null) {
+            return applySchemaless(record);
+        } else {
+            return applyWithSchema(record);
+        }
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public ConfigDef config() {
+        return CONFIG_DEF;
+    }
+
+    protected abstract Schema operatingSchema(R record);
+
+    protected abstract Object operatingValue(R record);
+
+    protected abstract R newRecord(R record, Schema updatedSchema, Object updatedValue);
+
+    private R applySchemaless(R record) {
+        final Map<String, Object> value = requireMap(operatingValue(record), PURPOSE);
+        final Map<String, Object> newValue = new LinkedHashMap<>();
+        applySchemaless(value, "", newValue);
+        return newRecord(record, null, newValue);
+    }
+
+    private void applySchemaless(Map<String, Object> originalRecord, String fieldNamePrefix, Map<String, Object> newRecord) {
+        for (Map.Entry<String, Object> entry : originalRecord.entrySet()) {
+            final String fieldName = fieldName(fieldNamePrefix, entry.getKey());
+            switch (ConnectSchema.schemaType(entry.getValue().getClass())) {
```

Comment:
Fair point. I've added handling for nulls that just applies the null to the same field and returns.